lib/utils/cobs.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2024 Kelly Helmut Lord
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/data/cobs.h>
+
+int cobs_encode(struct net_buf *src, struct net_buf *dst, uint32_t flags)
+{
+	int delimiter = COBS_FLAG_CUSTOM_DELIMITER(flags);
+
+	/* Calculate required space for worst case */
+	size_t max_encoded_size = src->len + (src->len / 254) + 2;
+
+	/* Check if destination has enough space */
+	if (net_buf_tailroom(dst) < max_encoded_size) {
+		return -ENOMEM;
+	}
+
+	uint8_t *code_ptr = net_buf_add(dst, 1);
+	uint8_t code = 1;
+
+	/* Process all input bytes */
+	uint8_t data = 0;
+
+	while (src->len > 0) {
+		data = net_buf_pull_u8(src);
+		if (data == delimiter) {
+			/* Delimiter found - write current code and start new block */
+			*code_ptr = code;
+			code_ptr = net_buf_add(dst, 1);
+			code = 1;
+		} else {
+			/* Add non-zero byte to output */
+			net_buf_add_u8(dst, data);
+			code++;
+
+			/* If we've reached maximum block size, start a new block */
+			if (code == 0xFF && (src->len - 1 >= 0)) {
+				*code_ptr = code;
+				code_ptr = net_buf_add(dst, 1);
+				code = 1;
+			}
+		}
+	}
+
+	*code_ptr = code;
+
+	if (flags & COBS_FLAG_TRAILING_DELIMITER) {
+		/* Add final delimiter */
+		net_buf_add_u8(dst, delimiter);
+	}
+
+	return 0;
+}
+
+int cobs_decode(struct net_buf *src, struct net_buf *dst, uint32_t flags)
+{
+	int delimiter = COBS_DEFAULT_DELIMITER;
```

Comment:
Flags always contains the delimiter.

```suggestion
	uint8_t delimiter = COBS_FLAG_CUSTOM_DELIMITER(flags);
```